### PR TITLE
Action to add calibrators from PanSTARRS

### PR DESCRIPTION
**Pull request by juanep97**
This PR adds an action to the AstroSource admin for automatically adding calibrators from PanSTARRS.

imagen

### CODE REVIEW

**Review by joteros — CHANGES_REQUESTED** [review]
Use ApMag instead of PSFMag
Set condition on the number of observations of each star, ApMagNpt>=5
Set condition on the ApMagStd<0.01

If with those conditions, we cannot find stars, then relax in a second step iteration those for ApMagNpt and ApMagStd

**Comment by juanep97** [comment]
@joteros what do you think of those constraints? I think they are a reasonable minimum. Then we just take the top 10 results, by number of observations.

**Comment by joteros** [comment]
Looks good to me. I usually use as minimum magnitude allowed 17 instead of 16, but this is a minor change. 
At some point we should add a second iteration search in case the search does not find any star in the FoV, relaxing a bit the criteria (it happened to me a couple of times), increasing the Std threshold, lowering the minimum magnitude and the number of observations required, but this can be done afterwards.

**Review by joteros — COMMENTED** [review]
Looks good to me. I usually use as minimum magnitude allowed 17 instead of 16, but this is a minor change.
At some point we should add a second iteration search in case the search does not find any star in the FoV, relaxing a bit the criteria (it happened to me a couple of times), increasing the Std threshold, lowering the minimum magnitude and the number of observations required, but this can be done afterwards.

File: iop4admin/modeladmins/astrosource.py [hunk 1]
```diff
@@ -74,3 +79,121 @@ def get_urls(self):
...
+                        'iMeanApMag', 'iMeanApMagErr', 'iMeanApMagStd', 'iMeanApMagNpt',
+                        'yMeanApMag', 'yMeanApMagErr', 'yMeanApMagStd', 'yMeanApMagNpt']
+        
+        for col in column_names:
+
+            idx = np.isfinite(catalog_data[col])
+            idx = idx.data & idx.mask
+            catalog_data = catalog_data[~idx]
+
+        idx = (11.0 <= catalog_data["rMeanApMag"]) & (catalog_data["rMeanApMag"] <= 16.0)
```
Comment (morcuended) [inline iop4admin/modeladmins/astrosource.py:115]: I would define these two magnitude minimum and maximum values at the beginning of the file. Or at least a comment explaining these cuts

File: iop4admin/modeladmins/astrosource.py [hunk 2]
```diff
@@ -74,3 +79,121 @@ def get_urls(self):
...
+        idx = idx & (catalog_data["rMeanApMagStd"] < 0.01)
+        idx = idx & (catalog_data["rMeanApMagNpt"] > 5)
```
Comment (morcuended) [inline iop4admin/modeladmins/astrosource.py:117]: what is `rMeanApMagNpt`, and why the value 5? (I would define it as well at the top)

File: iop4admin/modeladmins/astrosource.py [hunk 3]
```diff
@@ -74,3 +79,121 @@ def get_urls(self):
...
+
+        logger.info(f"Got {len(catalog_data)} PanSTARRS field stars around {main_src.name}")
+
+
+        column_names = ['objName', 'raMean', 'decMean',
+                        'rMeanApMag', 'rMeanApMagErr', 'rMeanApMagStd', 'rMeanApMagNpt',
+                        'gMeanApMag', 'gMeanApMagErr', 'gMeanApMagStd', 'gMeanApMagNpt',
+                        'iMeanApMag', 'iMeanApMagErr', 'iMeanApMagStd', 'iMeanApMagNpt',
+                        'yMeanApMag', 'yMeanApMagErr', 'yMeanApMagStd', 'yMeanApMagNpt']
+
```
Comment (morcuended) [inline iop4admin/modeladmins/astrosource.py:108]: Maybe add log info ("Filtering stars with magnitude between {MIN_R_MAGNITUDE} and {MAX_R_MAGNITUDE}")

File: iop4admin/modeladmins/astrosource.py [hunk 4]
```diff
@@ -74,3 +79,121 @@ def get_urls(self):
...
+        # sort by number of observations in R, take top 10 only
+        if len(catalog_data) > 10:
+            logger.info("Keeping only top 10 field stars by number of R observations")
+            catalog_data.sort('rMeanApMagNpt')
+            catalog_data = catalog_data[-10:]
+
+        field_stars = list()
+
+        for row in catalog_data:
+            try:
```
Comment (morcuended) [inline iop4admin/modeladmins/astrosource.py:142]: To make it clear I would add here a comment say also `# SDSS to Johnson-Cousins transformation by Lupton (2005)`

File: iop4admin/modeladmins/astrosource.py (hunk 2 above)
Comment (juanep97) [reply iop4admin/modeladmins/astrosource.py]: It is the number of observations for each aperture magnitude. I have defined the constraints at the top of the function and added some comments. The number 5 is arbitrary, but it is just the minimum. We take the top 10 results by this number.

File: iop4admin/modeladmins/astrosource.py (hunk 1 above)
Comment (juanep97) [reply iop4admin/modeladmins/astrosource.py]: Done.

File: iop4admin/modeladmins/astrosource.py (hunk 4 above)
Comment (juanep97) [reply iop4admin/modeladmins/astrosource.py:148]: Done

File: iop4admin/modeladmins/astrosource.py (hunk 3 above)
Comment (juanep97) [reply iop4admin/modeladmins/astrosource.py:114]: Done.

**Review by joteros — APPROVED** [review]

**Comment by juanep97** [comment]
> Looks good to me. I usually use as minimum magnitude allowed 17 instead of 16, but this is a minor change. At some point we should add a second iteration search in case the search does not find any star in the FoV, relaxing a bit the criteria (it happened to me a couple of times), increasing the Std threshold, lowering the minimum magnitude and the number of observations required, but this can be done afterwards.

I am hesitant to perform this automatic lowering of the quality constraints, since the user would not know about this and have no choice, at least through this GUI (if we added an interactive tab for it, perhaps). Another option would be to add two different actions, with two different set of constraints (e.g. "Automatically add field stars from PanSTARRS" and "Automatically add field stars from PanSTARRS (relaxed)").

**Comment by joteros** [comment]
> > Looks good to me. I usually use as minimum magnitude allowed 17 instead of 16, but this is a minor change. At some point we should add a second iteration search in case the search does not find any star in the FoV, relaxing a bit the criteria (it happened to me a couple of times), increasing the Std threshold, lowering the minimum magnitude and the number of observations required, but this can be done afterwards.
> 
> I am hesitant to perform this automatic lowering of the quality constraints, since the user would not know about this and would have no choice, at least through this GUI (if we added an interactive tab for it, perhaps). Another option would be to add two different actions, with two different sets of constraints (e.g. "Automatically add field stars from PanSTARRS" and "Automatically add field stars from PanSTARRS (relaxed)").

Sounds like a good solution. In any case, as I said this can be implemented (and even refined) afterwards, and my expectation is that it won't happen too often

**Pull request merged**